sklearn/cluster/optics_.py

```diff
@@ -527,3 +555,379 @@ def cluster_optics_dbscan(reachability, core_distances, ordering, eps=0.5):
+            U = _Area(start=U_start, end=end, maximum=reachability_plot[end],
+                      mib=-1)
+            mib = reachability_plot[end + 1]
+
+            U_clusters = []
+            for D in sdas:
+                c_start = D.start
+                c_end = min(U.end, n_samples - 1)
+
+                # line (**), sc2*
+                if reachability_plot[c_end + 1] * xi_complement < D.mib:
+                    continue
+
+                # Definition 11: criterion 4
+                if D.maximum * xi_complement >= reachability_plot[c_end + 1]:
+                    # Find the first index from the left side which is almost
+                    # at the same level as the end of the detected cluster.
+                    while (reachability_plot[c_start + 1] >
+                           reachability_plot[c_end + 1]
+                           and c_start < c_end):
+                        c_start += 1
+                elif reachability_plot[c_end] * xi_complement >= D.maximum:
+                    # Find the first index from the right side which is almost
+                    # at the same level as the beginning of the detected
+                    # cluster.
+                    while (reachability_plot[c_end] > D.maximum
+                           and c_end > c_start):
+                        c_end -= 1
+
+                # predecessor correction
+                c_start, c_end = _correct_predecessor(reachability_plot,
+                                                      predecessor, c_start,
+                                                      c_end)
+                if c_start is None:
+                    continue
+
+                # Definition 11: criterion 3.a
+                if c_end - c_start + 1 < min_cluster_size:
+                    continue
+
+                # Definition 11: criterion 1
+                if c_start > D.end:
+                    continue
+
+                # Definition 11: criterion 2
+                if c_end < U.start:
+                    continue
+
+                U_clusters.append((c_start, c_end))
+
+            # add smaller clusters first.
+            U_clusters.reverse()
+            clusters.extend(U_clusters)
+
+            if U.end == n_samples - 1:
+                break
+
+    return np.array(clusters)
+
+
+def _extract_xi_labels(ordering, clusters):
+    """Extracts the labels from the clusters returned by `_xi_cluster`.
+    We rely on the fact that clusters are stored
+    with the smaller clusters coming before the larger ones.
+
+    Parameters
+    ----------
+    ordering : array, shape (n_samples)
+        The ordering of points calculated by OPTICS
+
+    clusters : array, shape (n_clusters, 2)
+        List of clusters i.e. (start, end) tuples,
+        as returned by `_xi_cluster`.
+
+    Returns
+    -------
+    labels : array, shape (n_samples)
+    """
+
+    labels = np.zeros(len(ordering), dtype=np.int)
```

Comment:
Why not use `np.full(n_samples, -1, dtype=int)` as above, begin with `label = 0` in the next line, and drop the -1 operation below?